### PR TITLE
feat: don't request Solc JSON AST unless absolutely necessary

### PR DESCRIPTION
**Pull request by DaniPopes**
The Solc JSON AST output is very expensive to both serialize on the compiler's side, and deserialize on our side due to its very nested nature and generally big size.

This is especially noticeable on hot starts thanks to less time spent deserializing all the already-compiled artifacts. This is applicable to all commands that require a built project.

Adds the `ast = <bool>` config and `--ast` flag to build args to optionally include the AST in the artifacts.

We currently use the AST in:
- `coverage`
- `flatten`
- inline test config (`NatSpec`)

`coverage` and `flatten` have been updated to override the `ast` config.
`test` has not been updated for inline test config since it currently does not work, and all tests pass without it. It will have to be rewritten using `solang_parser` or something else.

## Results

Note that I'm using the debug profile to build `forge`, but this won't matter much as most of the time is spent in `solc` anyway.

Default `forge init` project:
image

Solmate:
image

sablier/v2-core:
- out/ size before: 144Mb
- out/ size after:  106Mb
- profile before: <https://share.firefox.dev/42KyHyM>
- profile after:  <https://share.firefox.dev/3T2RrpQ>
image

ethereum-optimism/optimism packages/contracts-bedrock
image

### CODE REVIEW

**Review by mattsse — COMMENTED** [review]
does this affect the regular forge build artifacts? 

~~we added metadata because this is apparently used by some tools~~

that's not part of metadata, so I think we can do that, but not entirely sure

**Review by mattsse — COMMENTED** [review]
I'd be fine with that.

we should check with slither first if they're relying on this in the artifacts file

**Review by onbjerg — CHANGES_REQUESTED** [review]
some external tools rely on this, so i'd add a flag to `forge build` to get this back

**Comment by mds1** [comment]
cc @0xalpharush for thoughts on how this affects slither

**Comment by gakonst** [comment]
Can we get a comparison for release mode compilation time improvements due to this on a large repo? e.g. Optimism contracts

**Review by onbjerg — APPROVED** [review]

File: crates/forge/bin/cmd/coverage.rs [hunk 1]
```diff
@@ -99,50 +102,42 @@ impl CoverageArgs {
...
-        };
+            // print warning message
+            p_println!(!self.opts.silent => "{}",
+            Paint::yellow(
+            concat!(
+            "Warning! \"--ir-minimum\" flag enables viaIR with minimum optimization, which can result in inaccurate source mappings.\n",
+            "Only use this flag as a workaround if you are experiencing \"stack too deep\" errors.\n",
+            "Note that \"viaIR\" is only available in Solidity 0.8.13 and above.\n",
+            "See more:\n",
+            "https://github.com/foundry-rs/foundry/issues/3357\n"
```
Comment (onbjerg) [inline crates/forge/bin/cmd/coverage.rs:126]: formatting lookig funky here

**Comment by DaniPopes** [comment]
> Can we get a comparison for release mode compilation time improvements due to this on a large repo? e.g. Optimism contracts

@gakonst
The overwhelming majority of the time is spent in solc so `release` mode on our side won't do much here.

It's 2-3% faster on clean builds, up to 5-10x for sablier/v2-core (absolute time save of about 2.5s for both). Note that time saved is proportional to amount of total artifacts.

**Comment by 0xalpharush** [comment]
I ran `foundryup -b dani/drop-ast` and it does make slither not work since the ASTs are missing. I think it is fine to drop from `forge build` but I'd prefer `forge build --build-info` (the hardhat artifact style) to be backwards compatible ofc. Otherwise, we'd have to cut a release and handle the case where people are not on this version of foundry yet

**Pull request merged**